### PR TITLE
[TRTLLM-13235][feat] Support bad_words in TorchSampler

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of blocked words during text generation, making sampling respect user-provided forbidden words more consistently.
  * Multi-token blocked phrases now suppress the final token only when the recent context matches, reducing unintended matches.
  * Added coverage to verify blocked-word behavior in sampling.

## Description

`TorchSampler` previously ignored `bad_words` / `bad_token_ids` (only the C++ `TRTLLMSampler` applied them). This PR adds enforcement, including correct behavior under the overlap scheduler.

### Bad-words masking in TorchSampler
- `llm_request.py`: mirror the executor request's `bad_words` as `py_bad_words`.
- `sampler.py`: add `TorchSampler._apply_bad_words` — single-token words are banned unconditionally; a multi-token word bans its final token only when the generated suffix matches the word prefix (matches C++ `ban_bad_words` semantics). Speculative steps reconstruct per-step context from `py_draft_tokens`.

### Overlap scheduler support
Under overlap, `sample_async(i)` runs before `update_requests(i-1)`, so the host token list lags the device by one token — multi-token prefix matching could miss or misapply bans. Fix, with no added device-to-host sync:
- Split the match: first `k-2` prefix tokens on the host, the newest token compared on-GPU against `store.new_tokens[0, seq_slot, 0]`; conditional ban applied via additive shape-static `index_put_`.
- Track per-slot pending (sampled-but-not-updated) steps to detect staleness; draft batches excluded.
- GPU path triggers only for overlap + no speculation + no beam search; spec/beam under overlap keeps prior behavior with a one-time warning. Non-overlap path unchanged.

### Post-merge review follow-ups for #16365 (@ixlmar)
- `ops/flashinfer.py`: missing flashinfer now raises an actionable `ImportError` on op use (module stays importable; startup enforcement stays in `TorchSampler.__init__`).
- `sampling_utils.py`: replace `import y as y` re-exports with plain imports + `__all__`; drop the `greedy` alias (callers use `greedy_search_sampling_batch`); `sanitize_top_k` uses `clamp(...).masked_fill_(...)`.

### Follow-up (out of scope)
Batched device-side bad-words matching via the stop-words machinery (`past_tokens_cuda`) would remove the host loop and cover spec/beam under overlap natively.

## Test Coverage
- `test_torch_sampler.py::TestApplyBadWords` (11 cases): regular path (single/multi-token hit/miss, prefix in prompt) and overlap path (device-side hit/miss, host+device combined, mixed stale/fresh batches, overlapping bans → no NaN).
- Full `test_torch_sampler.py` suite: 207 passed, no regressions (B200; skips are Ampere-gated).

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by zhaoyangwang-nvidia** [comment]
Hi @NVIDIA/trt-llm-runtime-devs please help to reivew this PR, only small feature, thanks~

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py [hunk 1]
```diff
@@ -4352,6 +4352,76 @@ def _apply_min_length_penalty(
...
+        """
+        if not any(getattr(r, "py_bad_words", None) for r in requests):
+            return logits
+
+        rows: list[int] = []
+        cols: list[int] = []
+        current_offset = 0
+        for index, r in enumerate(requests):
+            bad_words = getattr(r, "py_bad_words", None)
+            if bad_words:
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: Nit: `if not bad_words: continue` to avoid deep nesting

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py [hunk 2]
```diff
@@ -4352,6 +4352,76 @@ def _apply_min_length_penalty(
...
+                for beam_idx in range(num_beams[index]):
+                    # Already-generated tokens for this beam (excluding prompt).
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: If the prompt contains the start of a bad word, generation of said bad word would not be forbidden. Is this the desired behavior?

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py [hunk 3]
```diff
@@ -4695,6 +4765,13 @@ def _process_requests(
             sampling_requests_metadata.req_num_beams.tolist(),
         )
 
+        logits_cuda = self._apply_bad_words(
+            logits_cuda,
+            sampling_requests,
+            sampling_requests_metadata.req_num_steps.tolist(),
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: Move the `if not any(getattr(r, "py_bad_words", None) for r in requests)` check here, to avoid `.tolist()` if there are no bad words.

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py [hunk 4]
```diff
@@ -4352,6 +4352,76 @@ def _apply_min_length_penalty(
...
         return logits
 
+    @staticmethod
+    @torch.inference_mode()
+    def _apply_bad_words(
+        logits: torch.Tensor,
+        requests: list[LlmRequest],
+        num_steps: list[int],
+        num_beams: list[int],
+    ) -> torch.Tensor:
```
Comment (ixlmar) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: I'd suggest `-> None` to avoid confusion, since logits are modified in-place.

**Review by ixlmar — COMMENTED** [review]
@stnie Could this functionality be merged with the stop-words handling somehow?

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py (hunk 1 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: Fixed

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py (hunk 3 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: Fixed

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py (hunk 4 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: Fixed

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py (hunk 2 above)
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: Good catch, fixed.

**Comment by stnie** [comment]
> @stnie Could this functionality be merged with the stop-words handling somehow?

AFAIK stop words is applied after sampling, whereas bad words needs to be applied before sampling to adjust the logits.  One could try to reuse or adjust  `_are_stop_words` for batched processing of all bad words instead of the loop.

**Comment by zhaoyangwang-nvidia** [comment]
Hi @ixlmar This PR addresses your post-merge review comments on [#16365 (commit aded2c79)](https://github.com/NVIDIA/TensorRT-LLM/pull/16365):

- ops/flashinfer.py — no more silent late failures without flashinfer: calling any sampling op now raises an actionable ImportError with installation guidance instead of a bare NameError. The module intentionally stays importable without flashinfer (sampling_utils imports it unconditionally, and the non-flashinfer sampler paths must keep working); startup-time enforcement remains in TorchSampler.__init__, and the module docstring now documents that contract.
- sampling_utils.py — redundant re-exports cleaned up: the from x import y as y lines are collapsed into plain grouped imports plus an explicit __all__ (mypy runs in strict mode, so implicit re-export is disabled).
- greedy alias removed: drafting_loops.py and speculative/interface.py now use greedy_search_sampling_batch directly.
- sanitize_top_k simplified: torch.where(...).clamp(...) replaced with clamp(...).masked_fill_(...), avoiding the full_like intermediate; semantics are unchanged (verified against the old implementation for 0 / negative / INT32_MAX sentinels, input tensor not mutated).

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py [hunk 5]
```diff
@@ -3624,6 +3632,12 @@ def update_requests(
         if not state.requests:
             return
 
+        if self._track_pending_steps and not state.requests[0].py_is_draft:
```
Comment (lori-ren) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: L3635, L3801, L4834 check whether state.requests is draft from the first request, which assumes every request batch is homogeneous (all-draft or all-non-draft). Should we add some assertion here, or is there any test coverage?

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler.py]: Good catch —. Added _is_draft_batch() which asserts homogeneity and replaced all three call sites; a mixed batch now fails loudly instead of silently corrupting the pending-step counters.

File: tensorrt_llm/_torch/pyexecutor/sampler/sampler.py [hunk 6]
```diff
@@ -4650,6 +4845,37 @@ def _process_requests(
...
+                        # the missing host tokens cannot be reconstructed with
+                        # the single-token device-side check; multi-token bad
+                        # words may be applied one step late.
+                        self._warned_stale_bad_words = True
+                        logger.warning(
+                            "bad_words with the overlap scheduler and speculative "
+                            "decoding or beam search: multi-token bad words are "
+                            "matched against a host token history that lags the "
+                            "device state and may be enforced inexactly."
+                        )
```
Comment (lori-ren) [inline tensorrt_llm/_torch/pyexecutor/sampler/sampler.py:4868]: Perhaps `logger.warning_once` is more appropriate here and we can safely remove this flag

Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/sampler/sampler.py:4868]: Thanks，fixed。

**Review by lori-ren — APPROVED** [review]
